src/javascript/expression.rs

```diff
@@ -418,6 +448,139 @@ impl<'module> Generator<'module> {
+                ";",
+                line()
+            ]);
+            Ok(())
+        }
+        Pattern::Discard { .. } => Ok(()),
+        Pattern::Assign { name, pattern, .. } => {
+            assignments.push(docvec![
+                "let ",
+                name,
+                " = gleam$tmp",
+                concat(
+                    path.clone()
+                        .into_iter()
+                        .map(|i| i.clone().surround("[", "]"))
+                ),
+                ";",
+                line()
+            ]);
+            traverse_pattern(pattern, path, checks, assignments)
+        }
+
+        Pattern::List { elements, tail, .. } => {
+            let path_string = concat(path.into_iter().map(|i| i.clone().surround("[", "]")));
+            let length_check = match tail {
+                Some(_) => "?.length !== undefined".to_doc(),
+                None => "?.length === 0".to_doc(),
+            };
+            checks.push(docvec![
+                "gleam$tmp",
+                path_string,
+                Document::String("?.[1]".repeat(elements.len())),
+                length_check,
+            ]);
+            let _ = elements
+                .into_iter()
+                .enumerate()
+                .map(|x| {
+                    let (index, pattern) = x;
+                    let mut path = path.clone();
+                    for _ in 0..index {
+                        path.push("1".to_doc());
+                    }
+                    path.push("0".to_doc());
+
+                    traverse_pattern(pattern, &mut path, checks, assignments)
+                })
+                .collect::<Result<Vec<()>, Error>>()?;
+            if let Some(pattern) = tail {
+                let mut path = path.clone();
+                for _ in 0..elements.len() {
+                    path.push("1".to_doc());
+                }
+                traverse_pattern(pattern, &mut path, checks, assignments)?
+            }
+            Ok(())
+        }
+        Pattern::Tuple { elems, .. } => {
+            // We don't check the length, because type system means it's a tuple
+            let _ = elems
+                .into_iter()
+                .enumerate()
+                .map(|x| {
+                    let (index, pattern) = x;
+                    let mut path = path.clone();
+                    path.push(Document::String(format!("{}", index)));
+                    traverse_pattern(pattern, &mut path, checks, assignments)
+                })
+                .collect::<Result<Vec<()>, Error>>()?;
+            Ok(())
+        }
+        Pattern::Constructor {
+            constructor: PatternConstructor::Record { name },
+            ..
+        } if name == "True" => {
+            checks.push(equality(&path, "true".to_doc()));
+            Ok(())
+        }
+        Pattern::Constructor {
+            constructor: PatternConstructor::Record { name },
+            ..
+        } if name == "False" => {
```

Comment:
@lpil as far as I can see the PatternConstructor::Record is untyped the only varient is a record and the only field is name, therefore as the AST is now there is no way to check that the type's module is nil and therefore this could be a user type.

What I want to do is to implement `is_bool` `is_nil` somehow here, same as we use for expressions. Not sure which bit should be changed to add this.